### PR TITLE
MRG: Brainvision events_from_annotations

### LINKED ISSUE #6267: BrainVision annotation parsing is needlessly complicated
It used to be BV files would neatly map stim codes to integers: `'Stimulus/S 11'` would become `11` and so on. Now, with annotations, we have essentially random assignment; what `'S 11'` becomes is essentially unpredictable.

Now it is trivial to write a small function to get neater mappings, but 1. if you work with BV files, you will probably write that same function *every single time you import your data*, which is another 5 or so lines of essentially boilerplate, and 2. this presents a significant barrier of entry to newcomers.

Of course, it could be quite easy: least-surprise default mapping BV markers to integer codes in the way you'd expect, i.e., `'S 11'` becomes `11`.

Or am I missing something?

@cbrnr @massich

### PR DESCRIPTION
**Pull request by massich**
Closes #6267
Closes #6276

### CODE REVIEW

**Comment by larsoner** [comment]
Can you quickly say what's different relative to #6276?

**Comment by massich** [comment]
It simplifies the function and addresses some of the issues it had. But again this is the same as having a huge dict with the equivalences and return it.

File: mne/annotations.py [hunk 1]
```diff
@@ -856,6 +871,11 @@ def events_from_annotations(raw, event_id=None, regexp=None, use_rounding=True,
 
     annotations = raw.annotations
 
+    if event_id == 'auto' and hasattr(raw, 'auto_event_id'):
+        event_id = raw.auto_event_id()
```
Comment (larsoner) [inline mne/annotations.py]: It would be good to make this private rather than public as `_auto_event_id`

**Comment by larsoner** [comment]
@massich do you have more commits to push? If not, I can try and set it non-WIP

@jona-sassenhagen okay for you as requested by @cbrnr not to drop any of the events? I agree that just doing 2001, 2002, ... for all non-stim, non-resp BV markers in a `sorted` list would work. In my mind that's the one last unresolved issue. I can push a commit for this, too, if we agree it's useful (or @massich can, but it's post-work-hours in Europe now).

**Comment by jona-sassenhagen** [comment]
Yes ok with me.

**Comment by larsoner** [comment]
@jona-sassenhagen @cbrnr @massich can you look and make sure that you're happy with the docstring describing the behavior of this? I've also modified tests that hopefully verify it works according to the description.

**Comment by cbrnr** [comment]
The docstring looks good! If everything works as advertised, +1 for merge!

**Comment by larsoner** [comment]
Just pushed a commit to shorten the docstring, but the changes are minor

**Comment by massich** [comment]
@larsoner ok, I just saw what you did. The only fuzzy thing is why we need `lambda` on the default? 
here https://github.com/mne-tools/mne-python/pull/6326/files#diff-3552c91528439a66cb184c8d9b9cfcb5R873

is it 'cos we need something that returns `None` once called? I find it more tricky  to read but I like it.

**Comment by massich** [comment]
Shall we backport this one to 0.17? Will we be releasing the last 0.17.X?

**Comment by larsoner** [comment]
Usually after releasing a new version we stop backporting things, so no probably not. FYI here is the CircleCI rendering

https://13163-1301584-gh.circle-artifacts.com/0/dev/generated/mne.events_from_annotations.html

**Comment by larsoner** [comment]
> is it 'cos we need something that returns None once called? I find it more tricky to read but I like it.

Yes it's because we need the `()` at the end, it just saves us from a conditional branch (the conditional is implicit in the `get`)

**Comment by larsoner** [comment]
There is a merge conflict now, feel free to rebase

File: mne/io/brainvision/tests/test_brainvision.py [hunk 2]
```diff
@@ -462,11 +444,70 @@ def test_read_vmrk_annotations():
...
+    expected_orig_time = 1384359243.794231
+    expected_onset_latency = np.array(
+        [0, 486., 496., 1769., 1779., 3252., 3262., 4935., 4945., 5999., 6619.,
+         6629., 7629., 7699.]
+    )
+    expected_annot_description = [
+        'New Segment/', 'Stimulus/S253', 'Stimulus/S255', 'Stimulus/S254',
+        'Stimulus/S255', 'Stimulus/S254', 'Stimulus/S255', 'Stimulus/S253',
+        'Stimulus/S255', 'Response/R255', 'Stimulus/S254', 'Stimulus/S255',
+        'SyncStatus/Sync On', 'Optic/O  1'
```
Comment (jona-sassenhagen) [inline mne/io/brainvision/tests/test_brainvision.py]: Might also include an `S 10` in there to ensure it's parsed correctly.

**Comment by jona-sassenhagen** [comment]
LGTM, thanks @massich

File: mne/io/brainvision/tests/test_brainvision.py (hunk 2 above)
Comment (larsoner) [reply mne/io/brainvision/tests/test_brainvision.py]: you mean like https://github.com/mne-tools/mne-python/pull/6326/commits/0f1759e1d606b992aef1f5bed5fb8a2f7bb92f38 ?

**Comment by jona-sassenhagen** [comment]
Ok, now my issue with this is the following:

Iif I

- read BV files
- write the raw obj to disk as in fif format
- read this raw obj from disk
- extract events from annotations

it doesn't use the BV parser for my BV events, because the class of the read object is no longer RawBrainVision.

That is very counterintuitive I think.
Suggestions?
@massich @drammock

**Comment by agramfort** [comment]
then I don't see how we can "just make it work" out of the box.

you must accept to pass a specific parameter to the events_from_annotation
function.

>

**Comment by agramfort** [comment]
you can see if you can make the 'auto' mode smarter...

>

**Comment by larsoner** [comment]
We could at least allow passing `'bv'` or `'brainvision'` instead of `'auto'` to mean "use Brainvision mode". Then `events_from_annotaions(raw)` will not work but at least `events_from_annotations(raw, 'brainvision')` will.

**Comment by jona-sassenhagen** [comment]
That might be good, but to me the main issue is that it will confuse many people (who don't know how the background on our classes) that retrieving the events from essentially one and the same data object yields different results.

**Comment by larsoner** [comment]
> the main issue is that it will confuse many people (who don't know how the background on our classes) that retrieving the events from essentially one and the same data object yields different results.

True. If that consistency is more important than having it work with no arg for `RawBrainvision` classes, then we should get rid of `'auto'` and have people always do `'brainvision'`. So we have to choose if the consistency or convenience is more important here. To me the convenience (and backward compat / deprecation since this was in 0.18) makes it worth keeping `'auto'`.

And just to come back to what @agramfort mentioned, I don't see a way to easily have it work once you've save-loaded, unless we encode somewhere what the original type was. Candidates I can think of:

1. `raw.info['description']`, append something
        Not great because users could use this field.
2. `raw.annotations`, add some annotation of the original type
        Not great because then we're adding things that `events_to_annotations` will treat as events by default, users doing `raw.set_annotations` will lose this information, etc.

I can't immediately think of any other ways.

**Comment by jona-sassenhagen** [comment]
One very effortful way would be to have RawFIF objects created by saving a RawBrainVision object to also still belong to RawBrainvision ...

I'm really not sure what's the best solution here.

I think checking if all annotations conform to BV standards should work fairly well, and then a warning could recommend to use "bv" as the parser.
Generally, I think if all your events look like "Stimulus/S 1", "Stimulus/S20", then you probably want them mapped to 1, 20, ...

**Comment by larsoner** [comment]
> I think checking if all annotations conform to BV standards should work fairly well...
> Generally, I think if all your events look like "Stimulus/S 1", "Stimulus/S20", then you probably want them mapped to 1, 20, ...

If a user adds a single annotation on their own, or simply concatenates two RawBrainvision instances (which adds `BAD_boundary` annots) then what they get will change. This is inconsistent just in a different way (that involves common user usage patterns).

>  and then a warning could recommend to use "bv" as the parser

This part seems reasonable. I'd propose:

1. Keep current `'auto'` behavior class-based.
2. Add `'brainvision'` as a mode.
3. If the mode is `'auto'` and the input class `RawFIF` or `RawArray` and we find any markers that are BV-like, `warn` that they might want to use `event_id='brainvision'` instead of `event_id='auto'` (or use `mode=None` to avoid the warning)

**Comment by jona-sassenhagen** [comment]
>If a user adds a single annotation on their own, or simply concatenates two RawBrainvision instances (which adds BAD_boundary annots) then what they get will change. This is inconsistent just in a different way (that involves common user usage patterns).

I don't understand what you mean. Still the sensible way would be to link "Stimulus/S 20" to 20, etc.

What you specifically suggested makes sense to me.
Not volunteering :)

**Comment by larsoner** [comment]
> I don't understand what you mean. Still the sensible way would be to link "Stimulus/S 20" to 20, etc.

You said "checking if all" above -- I took this to mean a conditional "all" like `if all(bv_like(annot) for annot in annotations)`. "check if all" is too fragile for the reasons I mention above, you add a single annot on your own or just concat and the "check if all" will fail.

Did you mean "check if any"? This might also be okay, given that the BV events are pretty specifically formatted. The docstring (and behavior) would be updated to be something like:
```
        - **'auto' (default)**: prefer a raw-format-specific parser:
          - Brainvision: map ...

to 

        - **'auto' (default)**: prefer a specific parser:
          - If any annotations are Brainvision-like (e.g., "Stimulus/S 20"): map...
          - Otherwise: Behaves like None.
```
The only way I can think of that this will break is if someone adds on their own a BV-like annotation to their raw instance. But because the BV built-in annotations are pretty specifically formatted, I think this is probably an okay risk.

**Comment by jona-sassenhagen** [comment]
Ah right. Well, I originally thought you could have something like, if more than 3 and more than 90% of events are matched by the BV parser ...
But "check if any" might be preferable, as otherwise in non-auto mode nothing is guaranteed either right?

Or stick to warnings to be on the safe side.

**Comment by larsoner** [comment]
I slightly prefer the "any" route.

One other tricky part is if there is some BV user who adds their own (non-BV-like) annotations and does save-load round-trip and expects these to turn into events, 1) they will be surprised and 2) their code will not work the same way as in 0.18. But this is again a presumably pretty rare corner case.

**Comment by cbrnr** [comment]
I don't know, I would not expect a FIF file to be parsed like a BV file, even if we're just talking about events. I'd prefer `events_from_annotations(raw, 'brainvision')` to get back the desired events.

**Comment by larsoner** [comment]
> I don't know, I would not expect a FIF file to be parsed like a BV file, even if we're just talking about events. 

I agree for FIF files from the Neuromag scanner, but not for BV files that we have read in write to FIF format (mostly) just because it's the only output format we support. The problematic case that @jona-sassenhagen brings up seems to show the problem, basically:
```
raw = read_raw_brainvision(...)
raw.save(fname)
raw_read = read_raw_fif(fname)
events_from_annotations(raw_read) != events_from_annotations(raw)
```
This behavior is not so good.

In my mind the question is how big is the risk that someone using a non-BV file will add a BV-like event? Given the specificity of the event naming I think it's pretty low and we should just go to a `any-BV-like` parsing method.

**Comment by drammock** [comment]
+1 for the approach of looking at the annotations, and printing a warning if any of them look like a BV-style annotation and `mode != 'bv'`

candidate for backport?

**Comment by larsoner** [comment]
Yeah we'll want to backport. Sounds like people prefer to stick with the class-based check / [my original 3-step proposal above](https://github.com/mne-tools/mne-python/pull/6326#issuecomment-498693432).

@massich can you look?

**Comment by agramfort** [comment]
smarter "auto" and/or "brainvision" explicit param, either are fine with me.
Basically the parsing of the brainvision annotations is doing some black
magic
that only MNE does AFAIK. We pay a price for too magic but sometimes you
just want things to work...

>

**Comment by cbrnr** [comment]
@larsoner yes I think your approach is fine, but why do you need two parameters `mode` and `event_id`? Currently, `mne.events_from_annotations` only has an `event_id` parameter, which should suffice to implement what you suggest. Let's not add a new `mode` parameter if we don't really need it.

**Comment by agramfort** [comment]
sleeping over it. For the event_id='auto' as fif files don't have naturally
annotions we can then trigger
the check for brainvision style annotations (ie only do it for a RawFIF
instance)

**Comment by massich** [comment]
let me rephrase based on @larsoner snippet:

```py
raw = read_raw_brainvision(...)
raw.save(fname)
raw_read = read_raw_fif(fname)
events_from_annotations(raw_read, event_id='auto') != events_from_annotations(raw, event_id='auto')
```
and now we are proposing this:

```py

def events_from_annotations(raw, event_id="auto",
                            regexp=r'^(?![Bb][Aa][Dd]|[Ee][Dd][Gg][Ee]).*$',
                            use_rounding=True, chunk_duration=None,
                            verbose=None):
...
    event_id = _check_event_id(event_id, raw)
...


def _check_event_id(event_id, raw):
    from mne.io.brainvision import _BVEventParser as bv_default
    from mne.io.brainvision import _check_raw_fif_from_original_bv as black_magic

    if event_id is None:
        return _DefaultEventParser()
    elif event_id == 'brainvision':
        return bv_default()
    elif event_id == 'auto':
        candidate = getattr(raw, '_get_auto_event_id', _DefaultEventParser)()

        if black_magic(raw) and candidate != bv_default:
            warn("you might want to try `event_id='brainvision'`)

        return candidate
    elif callable(event_id) or isinstance(event_id, dict):
        return event_id
    else:
        raise ValueError('Invalid input event_id')
```

So that this would happen:
```py
raw = read_raw_brainvision(...)
events_when_raw_was_bv = events_from_annotations(raw)
raw.save(fname)
raw_read = read_raw_fif(fname)
events_from_annotations(raw_read, event_id='auto') != events_when_raw_was_bv
# you get a warining

events_from_annotations(raw_read, event_id='brainvision') == events_when_raw_was_bv
```

My opinion is that `'auto'` is a bad idea. I think that `events_from_annotations`
requires `event_id` as much as requires `raw`. We cannot get away with
`events_from_annotations(raw.annotations)` because we need `raw.sfreq`. In the same
manner we cannot assign `ids` if we don't know them. I've never seen a problem with
a code like this:

```py
from mne.io.brainvision import event_ids as bv_events
raw = read_raw_fif(fname_of_an_old_bv)
events, _ = events_from_annotations(raw, event_id=bv_events)
```

If you want to get brainvision events (or any other type) you need to pass a proper
`event_id` that translate to those. I don't see this as a complicate and obscure coding
pattern for newcomers (We have much worst).

Educating the users that you need `event_id` or you get  `_DefaultEventParser` which gives
you consecutive numeric ids for each description is not the end of the world. Otherwise,
we update `_DefaultEventParser` to behave like `mne.io.brainvision.bv_events` as it was
in 0.16 and `_DefaultEventParser` would start giving you `ids` starting at (3000).

I could see people arguing that using strings is better (maybe, I'm not sure. I like the idea of having to import the translator. I find it more explicit):
```py
raw = read_raw_fif(fname_of_an_old_bv)
events, _ = events_from_annotations(raw, event_id='brainvision')
```

But I fear that if we start introspecting and guessing which was the original format
we are opening a can of worms.

At this point I would just make `'auto'` behave like `None` and deprecate it (because we released in 0.18).
And if we want `'brainvision'` and `'bla'` to be shorthands to avoid the imports thats fine with me.

**Comment by jona-sassenhagen** [comment]
>But I fear that if we start introspecting and guessing which was the original format we are opening a can of worms.

In prickle yes, but in practice BV stimulus events are extremely regular and specific, and everyone who uses events like "Stimulus/S100" will want it to be parsed to `100`.

>I've never seen a problem with a code like this:

Again, in principle yes, but right now a lot of people are working with `mne.find_events` still, some are making the transition right now, others might mostly use other formats than BV and so on.

It seems to me like the point of annotations is to make working with non-int events easier, but here we have something that the experimenter has programmed into their software as an int, is stored by BV as an int plus some predictable boiler plate, and is then universally parsed as an int. *But we somehow make it hard to use?* It's literally 8-bit integer codes!

To be frank, this should have been handled back when annotations were conceived ... I should have been more involved there, to represent BV. Now we will end up with a less than optimal solution either way.


>At this point I would just make 'auto' behave like None and deprecate it (because we released in 0.18).

Ok, maybe? If we add big warning messages for every usage of events_from_annotations on RawBrainvision objs for a cycle or three, and tell every BV user to always use event_id='brain vision' in all BV-derived code.

**Comment by massich** [comment]
```py
def _check_event_id(event_id, raw):
    from mne.io.brainvision import _BVEventParser as bv_default

    if event_id is None:
        return getattr(raw, '_get_auto_event_id', _DefaultEventParser)()
    elif event_id == 'auto':
        warn("`event_id='auto'` is deprecated. Use  `event_id=None` instead",
              DeprecationWarning)
        return getattr(raw, '_get_auto_event_id', _DefaultEventParser)()
    elif event_id == 'brainvision':
        return bv_default()
    elif callable(event_id) or isinstance(event_id, dict):
        return event_id
    else:
        raise ValueError('Invalid input event_id')
```

**Comment by larsoner** [comment]
> elif event_id == 'auto':
> <deprecate>

-1 on this. I am with @jona-sassenhagen that this mode is useful for BV users and we can potentially extend it to others (e.g., CTF once it's done).

The proposal that @agramfort seems to support now is actually not the one above (about warning) but rather a modified version of [the second proposal](https://github.com/mne-tools/mne-python/pull/6326#issuecomment-498697193), which I think would be:

- instance of RawBrainvision, use BV parser
- instance of RawFIF with any BV-like event, use BV parser
- (also probably want: instance of RawArray with any BV-like event, use BV parser)

I'm fine with this. It does not seem too fragile or risky, and simplifies things for BV users.

This can also be how we treat other readers. For example for the CTF annotations we are adding, if they end up having standard names -- instance of RawFIF/RawArray with CTF-like annotations, or RawCTF: use CTF event parsing. So it's a sustainable approach, too. (And in extreme corner cases where CTF and BV-like events are found, which should be pretty much never, we can raise an error or warn or something.)

**Comment by massich** [comment]
> To be frank, this should have been handled back when annotations were conceived ... I should have been more involved there, to represent BV. Now we will end up with a less than optimal solution either way.

It's not a problem of the annotation changes. Before 0.15 events_from_annotations was not general, had 'brainvision' logic embedded in it (which was not even complete. because if you had `'Foo/F100'` and `'Bar/B100'` both were mapped to 100). The only thing we did was make it explicit. You need to pass a dict, callable or anything that makes the events at your liking. What its true is that we changed the default from a half-implemented `'brainvision'` specification to something that was giving incremental numbers. If we want to make the `'brainvision'` specification the default just because this is the main use case for 99.99% of our users we can argue that. But again, the events_id would not be the same as in 0.15 because now we have a more complete implementation of the specification.

Despite having the opinion that:
-  'brainvision' is only 1 of our 15 readers and thats why it should not be the default. The default should be something really simple and let every reader expose something to translate. 
I will stay out of the pull. I'll be **+0** in making 'barinvision' the default of `events_from_annotations`.

**Comment by massich** [comment]
@larsoner 	what are you -1 ? to the warning? we could keep both approaches:
```py
def _check_event_id(event_id, raw):
    from mne.io.brainvision import _BVEventParser as bv_default

    if event_id is None or event_id='auto':
        return getattr(raw, '_get_auto_event_id', _DefaultEventParser)()
    elif event_id == 'brainvision':
        return bv_default()
    elif callable(event_id) or isinstance(event_id, dict):
        return event_id
    else:
        raise ValueError('Invalid input event_id')
```

**Comment by cbrnr** [comment]
This discussion shows that the underlying problem is really that we now have two things to represent discrete events - see #6018. Whatever we do here, it will never be perfect, because we decided to represent discrete events stored in files as annotations. Annotations contain everything you need, including string descriptions. I know that superficially it is easier to work with a NumPy array, but you could do everything with annotations. We'd only need to add support for annotations everywhere we currently accept events.

**Comment by larsoner** [comment]
> what are you -1 ? to the warning?

I am -1 to deprecating `'auto'`. I am +1 to what @agramfort came around to, which is the proposal to check BV class, then check if RawFIF or RawArray with bv-like events. It seems cleanest to me.

> We'd only need to add support for annotations everywhere we currently accept events.

Yes in this case the "only" includes critically `mne.Epochs`, and really we'd just end up with the same problem there (how to auto-create `event_id` to extract BV events). But let's keep discussion of this stuff contained to #6018 so we do not get derailed here.

**Comment by cbrnr** [comment]
> Yes in this case the "only" includes critically mne.Epochs, and really we'd just end up with the same problem there (how to auto-create event_id to extract BV events). But let's keep discussion of this stuff contained to #6018 so we do not get derailed here.

Yes, let's discuss in #6018 - but in the case of Epochs and annotations, why would you need event IDs when the user would specify the string descriptions that should be used for epoch creation?

**Comment by massich** [comment]
> Yes, let's discuss in #6018 - but in the case of Epochs and annotations, why would you need event IDs when the user would specify the string descriptions that should be used for epoch creation?

```py
raw = read_raw_brainvision(...)
raw.save(fname)
raw_read = read_raw_fif(fname)

events, event_id = events_from_annotations(raw)
# or any other of those:
#
# events, event_id = events_from_annotations(raw, event_id='auto')
# events, event_id = events_from_annotations(raw, event_id='brainvision')
# event_id = mne.io.brainvision.bla
# events, _ = events_from_annotations(raw, event_id=event_id)

epoch = Epochs(raw, events=events, event_id=event_id)
```

The thing is that until now `Epochs` require `events` and `event_id` to make the link.
Because all this `events_from_annotations` that takes and returns `event_id` is to call
Epochs.
I agree with @cbrnr that if this was possible, then there would be no discussion

```py
raw = read_raw_brainvision(...)
raw.save(fname)
raw_read = read_raw_fif(fname)
my_descriptors = ['foo', 'bar']
epoch = Epochs(raw, my_descriptors)
```

**Comment by massich** [comment]
The other thing that nobody mentioned is the elephant in the room: **Why everything has to be fif**. If I could save brainvision files I would have no trouble.

```py
raw = read_raw_brainvision('name.vhdr')
raw.save('bla.vhdr')
raw2 = read_raw_brainvision('bla.vhdr')
```
and then suport `read_raw`:

```py
raw = read_raw('name.vhdr')
raw.save('bla.vhdr')
raw2 = read_raw('bla.vhdr')
```

To me an image is an image, regardless of the format that once it was stored.

**Comment by agramfort** [comment]
what I propose has minimal consequences on project API and makes it work
for Jona....

**Comment by massich** [comment]
@agramfort there's a small PR comming with those changes. To discuss in concrete terms.

**Comment by jona-sassenhagen** [comment]
I’d think alex’ second proposal should work. 

A BV writer I don’t like. .fif isn’t perfect, but a lot of work has gone into making it a standard. And you’ll save your epochs and Ica Objs as fif too.
More generally, instead of writing a raw BV using MNE, you could just stick to using the original raw file ...

**Comment by jona-sassenhagen** [comment]
Great thanks @massich

**Comment by cbrnr** [comment]
> I agree with @cbrnr that if this was possible, then there would be no discussion

```
raw = read_raw_brainvision(...)
raw.save(fname)
raw_read = read_raw_fif(fname)
my_descriptors = ['foo', 'bar']
epoch = Epochs(raw, my_descriptors)
```

And why shouldn't this be possible? We could check whether the second argument is an nd-array and treat it as events or a string/list of strings and get the onsets from `raw.annotations`. Doesn't look too difficult, but maybe I'm missing something (because I've never really used `Epochs`).

Re the elephant, I agree that FIFF is not a universal format. Few tools outside of MNE can even read FIFF, so if I really want to store some data I go for something else, e.g. the BIDS-EEG supported formats EDF or BrainVision, or even EEGLAB, which is just a MAT file many tools can read. (For the sake of reproducibility, it's always best to only use the raw data, but I know that there are situations where you want to store some intermediate results to disk.)

Currently, MNE can't export any of these three formats for different reasons, so I've started to implement that in MNELAB. Currently, I can export to EEGLAB and EDF/BDF, and I will add [BrainVision export](https://gitlab.com/palday/philistine) based on the implementation by @palday soon. We could discuss if any of these formats might be suitable for direct integration into MNE.

Re `read_raw`, this discussion comes up regularly, e.g. #1930. I'd also be +1 for having a generic reader function, but it seems to be quite a lot of effort.

**Comment by agramfort** [comment]
>
> And why shouldn't this be possible? We could check whether the second argument is an nd-array and treat it as events or a string/list of strings and get the onsets from raw.annotations. Doesn't look too difficult, but maybe I'm missing something (because I've never really used Epochs).
>
>
that could work but a number of functions that take epochs as input will
use the epochs.events attribute. We need to be careful that certain epochs
will not work for certain functions. I am afraid of a rabbit hole here.

**Comment by cbrnr** [comment]
> that could work but a number of functions that take epochs as input will
> use the epochs.events attribute. We need to be careful that certain epochs
> will not work for certain functions. I am afraid of a rabbit hole here.

But we could start with adding this feature while not touching the old behavior at all (i.e. leaving the events-based interface as default).

**Comment by massich** [comment]
I'm in favor of all those things. But we need a discussion and a roadmap.

**Comment by agramfort** [comment]
events_from_annotations is one more line of code and keeps the magic
level one step lower
but I am open to discussion to see what we can cook up

**Comment by hoechenberger** [comment]
> Re the elephant, I agree that FIFF is not a universal format. Few tools outside of MNE can even read FIFF, so if I really want to store some data I go for something else, e.g. the BIDS-EEG supported formats EDF or BrainVision, or even EEGLAB, which is just a MAT file many tools can read. (For the sake of reproducibility, it's always best to only use the raw data, but I know that there are situations where you want to store some intermediate results to disk.)

This has actually been a hug pain for me in the past too. None if my non-MNE-using colleagues can read those FIFF files, let alone know which software could _potentially_ be used to open them.
